NXP backend: Resolve limitations of uncertain tensor formats.

MartinPavella opened the pull request:
### Summary
This PR resolves format related issues by inferring the format (NCHW/NHWC) for all nodes before partitioning. These formats are then used by the NeutronPartitioner to accurately determine which nodes are supported on Neutron.

### Test plan
Unit tests provided, and correct function is tested by nearly every test in the nxp backend.


cc @kimishpatel

MartinPavella committed:
NXP backend: Store inferred node format in the `node.meta`.

MartinPavella committed:
NXP backend: Update pass which removes GetItem nodes, to preserve the node format.

The pass `RemoveGetItemPass` replaces a `max_pool2d_with_indices` node with a `max_pool2d` node, that doesn't require a GetItem afterward. The new operator must, however, preserve the original node format. Therefore, a copy of the pass was created in `backends/nxp/_passes`, where it was modified. The new directory was created, because the pass doesn't follow the `NeutronEdgePass` interface.

MartinPavella committed:
NXP backend: Perform node format inference before partitioning.

Before, the format inference was done during conversion to NeutronIR (after partitioning), so the partitioner didn't yet know the formats. Now, the partitioner has the format data, which can be used to accurately select nodes for delegation.

MartinPavella committed:
NXP backend: Improve `cat` delegation by using inferred node formats.

MartinPavella committed:
NXP backend: Improve `constant_pad_nd` delegation by using inferred node formats.

MartinPavella committed:
NXP backend: Do not infer format for unknown nodes.

MartinPavella commented:
@pytorchbot label "module: nxp" "release notes: nxp"

kimishpatel commented:
i have imported it internally. There is an nxp test failing as well

digantdesai reviewed (approved):
Review automatically exported from Phabricator review in Meta.

MartinPavella commented:
I have fixed the failing neutron tests, and in the newest commit I have added the neutron backend build to the `unittest-buck2.sh`. @kimishpatel can you please verify I did it correctly? An can you confirm that the internal Meta tests are passing now?

MartinPavella commented:
@kimishpatel I have fixed the failing nxp test, and added the nxp backend build to the `unittest-buck.sh`. The build is now failing with:
1: From load at backends/nxp/TARGETS:4
2: File not found: `shim_et//arvr/tools/build_defs/oxx_python.bzl`
Do you know what is causing the issue?

kimishpatel commented:
> @kimishpatel I have fixed the failing nxp test, and added the nxp backend build to the `unittest-buck.sh`. The build is now failing with:
> 
> 1: From load at backends/nxp/TARGETS:4 2: File not found: `shim_et//arvr/tools/build_defs/oxx_python.bzl`
> 
> These lines were in the file already before my changes. Do you know what is causing the issue/how to fix it?

I guess it is triggering your build in oss and we have never really had this test in oss so its trying to pull in internal files. I would say roll back that change for now just to be able to land.

MartinPavella commented:
> > @kimishpatel I have fixed the failing nxp test, and added the nxp backend build to the `unittest-buck.sh`. The build is now failing with:
> > 1: From load at backends/nxp/TARGETS:4 2: File not found: `shim_et//arvr/tools/build_defs/oxx_python.bzl`
> > These lines were in the file already before my changes. Do you know what is causing the issue/how to fix it?
> 
> I guess it is triggering your build in oss and we have never really had this test in oss so its trying to pull in internal files. I would say roll back that change for now just to be able to land.

@kimishpatel thanks for the response. I have removed the build as you said. Can you please now verify that your internal build passes?

StrycekSimon reviewed (changes requested):
Please update the `numMacs` related comment. Did the `remove_geititem_pass.py` just pass the linter with such long lines?

MartinPavella commented:
> > > doing
> > 
> > 
> > @kimishpatel any updates? Is the build passing?
> 
> @kimishpatel pinging in case you missed this.

@kimishpatel any updates? This issue is blocking multiple other PRs. I really don't want to break your internal builds again, so it would be great to confirm they are passing before merging this.

kimishpatel commented:
Sorry for the late response on this. Lets rebase and then merge it
